test/gtest/common/test_helpers.cc

```diff
@@ -15,6 +15,133 @@ namespace ucs {
 
 const double test_timeout_in_sec = 60.;
 
+const double watchdog_timeout_default = 900.; // 15 minutes
+
+static test_watchdog_t watchdog;
+
+void *watchdog_func(void *arg)
+{
+    int ret = 0;
+    double now;
+    struct timespec timeout;
+
+    pthread_mutex_lock(&watchdog.mutex);
+
+    /* sync with parent thread */
+    pthread_barrier_wait(&watchdog.barrier);
+
+    do {
+        now = ucs_get_accurate_time();
+        ucs_sec_to_timespec(now + watchdog.timeout, &timeout);
+
+        ret = pthread_cond_timedwait(&watchdog.cv, &watchdog.mutex, &timeout);
+        if (!ret) {
+            switch (watchdog.state) {
+            case test_watchdog_t::WATCHDOG_RUN:
+                if (watchdog.timeout != watchdog_timeout_default) {
+                    /* reset to the default one */
+                    watchdog.timeout = watchdog_timeout_default;
+                }
+                break;
+            case test_watchdog_t::WATCHDOG_STOP:
+                /* force the end of the cycle */
+                ret = 1;
+                break;
+            case test_watchdog_t::WATCHDOG_TIMEOUT_SET:
+                /* this value will be reseted to the default one when signal
+                 * will be received when the test completed */
+                watchdog.state = test_watchdog_t::WATCHDOG_RUN;
+                break;
+            }
+            pthread_barrier_wait(&watchdog.barrier);
+        }
+    } while (!ret);
+
+    if (ret && (watchdog.state != test_watchdog_t::WATCHDOG_STOP)) {
+        if (ret == ETIMEDOUT) {
+            ADD_FAILURE() << "Timeout expired - abort";
+            pthread_kill(watchdog.parent_thread, SIGABRT);
+        } else {
+            ADD_FAILURE() << "Watchdog fails " << ret;
+            abort();
+        }
+    }
+
+    pthread_mutex_unlock(&watchdog.mutex);
+
+    return NULL;
+}
+
+void watchdog_signal(int barrier)
+{
+    pthread_mutex_lock(&watchdog.mutex);
+    pthread_cond_signal(&watchdog.cv);
+    pthread_mutex_unlock(&watchdog.mutex);
+
+    if (barrier) {
+        pthread_barrier_wait(&watchdog.barrier);
+    }
+}
+
+void watchdog_timeout_set(double new_timeout)
+{
+    pthread_mutex_lock(&watchdog.mutex);
+    /* change timeout value */
+    watchdog.timeout = new_timeout;
+    watchdog.state   = test_watchdog_t::WATCHDOG_TIMEOUT_SET;
+    /* apply new value for timeout */
+    watchdog_signal(0);
+    pthread_mutex_unlock(&watchdog.mutex);
+
+    pthread_barrier_wait(&watchdog.barrier);
+}
+
+void watchdog_start()
+{
+    pthread_mutexattr_t mutex_attr;
+
+    pthread_mutexattr_init(&mutex_attr);
+    /* create reentrant mutex */
+    pthread_mutexattr_settype(&mutex_attr, PTHREAD_MUTEX_RECURSIVE);
```

Comment:
where is recursiveness used?

Reply:
here:
https://github.com/openucx/ucx/blob/7975cb40c961daee9da16a262216bf5ef749f9bb/test/gtest/common/test_helpers.cc#L93

Do you this is not a good idea? I can avoid it by using `pthread_cond_signal` w/o wrapper